alpha/veneer/semver/semver.go

```diff
@@ -0,0 +1,428 @@
+)
+
+// mapping channel name --> priority, where greater-is-better
+var channelPriorities = map[string]int{CandidateChannelName: 0, FastChannelName: 1, StableChannelName: 2}
+
+// sorting capability for a slice according to the assigned channelPriorities
+type byChannelPriority []string
+
+func (b byChannelPriority) Len() int { return len(b) }
+func (b byChannelPriority) Less(i, j int) bool {
+	return channelPriorities[b[i]] < channelPriorities[b[j]]
+}
+func (b byChannelPriority) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+
+// map of channels : bundles : bundle-version
+// channels --> bundles --> version
+type semverRenderedChannelVersions map[string]map[string]semver.Version // e.g. d["stable-v1"]["example-operator/v1.0.0"] = 1.0.0
+
+func (v Veneer) Render(ctx context.Context) (*declcfg.DeclarativeConfig, error) {
+	var out declcfg.DeclarativeConfig
+
+	sv, err := ReadFile(v.Ref)
+	if err != nil {
+		log.Fatalf("semver-render: unable to read file: %v", err)
+	}
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, b := range sv.Candidate.Bundles {
+		r := action.Render{
+			AllowedRefMask: action.RefBundleImage,
+			Refs:           []string{b.Image},
+		}
+		c, err := r.Run(ctx)
+		if err != nil {
+			return nil, err
+		}
+		cfgs = append(cfgs, *c)
+	}
+	out = *combineConfigs(cfgs)
+
+	channelBundleVersions, err := sv.getVersionsFromStandardChannels(&out)
+	if err != nil {
+		log.Fatalf("semver-render: unable to post-process bundle info: %v", err)
+	}
+
+	channels := sv.generateChannels(channelBundleVersions)
+	out.Channels = channels
+	out.Packages[0].DefaultChannel = sv.defaultChannel
+
+	return &out, nil
+}
+
+func ReadFile(ref string) (*semverVeneer, error) {
+	data, err := ioutil.ReadFile(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	// default behavior is to generate only minor channels and to use skips over replaces
+	sv := semverVeneer{
+		GenerateMajorChannels: false,
+		GenerateMinorChannels: true,
+		AvoidSkipPatch:        false,
+	}
+	if err := yaml.Unmarshal(data, &sv); err != nil {
+		return nil, err
+	}
+	return &sv, nil
+}
+
+func (sv *semverVeneer) getVersionsFromStandardChannels(cfg *declcfg.DeclarativeConfig) (*semverRenderedChannelVersions, error) {
+	versions := semverRenderedChannelVersions{}
+
+	bdm, err := sv.getVersionsFromChannel(sv.Candidate.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[CandidateChannelName] = bdm
+
+	bdm, err = sv.getVersionsFromChannel(sv.Fast.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[FastChannelName] = bdm
+
+	bdm, err = sv.getVersionsFromChannel(sv.Stable.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[StableChannelName] = bdm
+
+	return &versions, nil
+}
+
+func (sv *semverVeneer) getVersionsFromChannel(semverBundles []semverVeneerBundleEntry, cfg *declcfg.DeclarativeConfig) (map[string]semver.Version, error) {
+	entries := make(map[string]semver.Version)
+
+	for _, semverBundle := range semverBundles {
```

Comment:
Do we even need `semverBundles` here? I wonder if we can just iterate `cfg.Bundles` to build the map of `cfg.Bundles[i].Image` to `cfg.Bundles[i].Properties[?(type == olm.package)].Version`?